### PR TITLE
Fix ppc64 TOC load for exception handler addresses

### PR DESCRIPTION
**Pull request by mshinwell**
GPR #2237 contains a bug which was noticed by precheck failing on GPR #2268.  Interestingly, precheck only failed on the flambda build on ppc64le, but not on the non-flambda build.

The reason for the failure is a bogus exception handler address in a trap frame.  This is on ppc64le with the ELF64v2 ABI, an ABI of impressive complexity, of which I understand very little.  On this system an address amidst the code (such as that of an exception handler) is obtained by labelling the relevant code and putting a reference to that label in a special place in the data section called the "table of contents".  To load the code address at runtime, the base address of the table will be found (always in register `r2`), and a signed integer constant will be added to it.  This will then point at the slot in the table that in turn contains the code address.

The computation to load an exception handler address is therefore of the form: `r0 = *(r2 + C)`, where `r0` is a temporary that we are going to store in the trap frame, and `C` is the integer constant.  When we emit the assembly we don't have to determine what `C` is; instead, we label the slot in the table of contents and use special linker support to substitute the label reference with `C`, which can be determined at link time.

Owing to constraints on the sizes of constants that can be fitted into the relevant instructions, the computation is actually going to be split into `r0 = *(r2 + A + B)`, in certain cases.  One of these cases is when `-for-pack` is in effect (@xavierleroy perhaps you could explain why this is) and is the reason that GPR#2227 did not cause CI to fail.  At present, there are no native code packs in the compiler distribution; GPR#2268 will introduce the first one.

In order to compute `r0 = *(r2 + A + B)` we generate two instructions: `r0 = r2 + A` then `r0 = *(r0 + B)`.  The second instruction in ppc64 assembly language is of the form `ld 0, exp(0)`, with `exp` being the appropriate TOC slot label reference and relocation that I shall elide.  The linker will turn `exp` into `B`.  (For some strange reason, you elide the `r` prefix when referencing registers in written ppc64 assembly code.)

In the test case that failed, B was computed by the linker to be some negative number, let's say -16000.  Upon disassembling the test case I was surprised to see that the `ld` instruction didn't say `ld`, it said `li`, which was just loading a constant (-16000, or whatever) into `r0`.   This address was the one in the trap frame, which was obviously wrong.

This turns out to be because writing zero between the brackets of an `ld` instruction doesn't mean _register_ zero (unlike writing the number one, or two, for example).  It means zero!  Hence why `ld` had been turned into `li`.

Why didn't this fail on the non-flambda build?  The reason appears to be because, going back to the notation above, `A` was zero.  This presumably happens because the code is smaller due to less inlining without flambda.  As such, the linker actually optimised the sequence, replacing the first `add` instruction with a no-op and the second (`ld`) instruction with one that just added `B` to the contents of `r2` before dereferencing.  The strange thing here is that this optimisation appears to be invalid given the semantics of "zero between the brackets".  Our (broken) input code described a computation that loaded a constant integer into a register, but the output code miraculously loaded something via the TOC.  Magic.

The fix is not to use `r0` as the temporary.

### CODE REVIEW

**Comment by dra27** [comment]
Rather than relying on #2268 to exercise this, can we contrive a smaller test case?

File: asmcomp/power/emit.mlp [hunk 1]
```diff
@@ -198,7 +198,12 @@ let emit_toctable () =
       `{emit_label lbl}:	.quad	{emit_tocentry entry}\n`)
     tocref_entries
 
-(* Emit a load from a TOC entry *)
+(* Emit a load from a TOC entry.
+
+   This function must not be called with any destination that resolves to
+   register r0.  A zero operand between the brackets of an "ld"
```
Comment (xavierleroy) [inline asmcomp/power/emit.mlp]: The comment reads off as a bit naive, if I may :-(  I would have written:
```
The destination is used as index register for a ld instruction, but r0 reads as zero when used as an index register.
```

**Review by xavierleroy — APPROVED** [review]
This is a good fix.  Just splitting hair on a comment below, but feel free to ignore.

**Comment by xavierleroy** [comment]
To give a bit of background:

Yes, register 0 is special in the Power ISA.  For some instructions it reads as zero, e.g. when used as index register in load and store instructions; also when used as source register in `addi` and `addis`, IIRC.  For other instructions register 0 behaves like a normal register.

The default syntax for Power assembly language is indeed to write registers as `0`, `1`, ..., `31` so that they cannot be confused with symbols.  The context (the instruction used) disambiguates between register number and integer literal, e.g.
```
xor 0, 0, 0    // r0 <- r0 ^ r0
xori 0, 0, 0   // r0 <- r0 ^ 0
```
It is a nightmare to read, indeed.

The GNU assembler has an option `-mregnames` to use `r0`...`r31` as register names.  With this option, asm code becomes a lot more legible:
```
xor r0, r0, r0    // r0 <- r0 ^ r0
xori r0, r0, 0   // r0 <- r0 ^ 0
```
Plus, the assembler will complain if you write `r0` in a context where register 0 reads as zero, forcing you to write `0` instead:
```
addi r0, r0, 12  // error!
addi r0, 0, 12   // good!
```
That would have found the bug earlier, as an assembly-time error.  On the other hand, it takes a little more work in configure to select the `-mregnames` option for Power if GNU tools are used.

**Comment by xavierleroy** [comment]
@dra27: a smaller test case could be obtained by passing the `-flarge-toc` option to ocamlopt (but it is specific to the Power target).

**Comment by mshinwell** [comment]
@xavierleroy Thanks for the updated comment and the extra information.  Whilst we are at it, could you clarify why `-for-pack` always uses the big-TOC load sequence?

I've pushed a changeset to update the comment and add a testcase.  I don't think the testcase is 100% certain to fail without this fix, although it does fail on the machine I tried it on.

**Comment by xavierleroy** [comment]
> why -for-pack always uses the big-TOC load sequence?

My memory is hazy.  But I think the reasoning is as follows.  When several object files are packed, they end up sharing the same TOC.  (Each .o file has a single TOC, even though the final program can have several.)  The resulting combined TOC is often too big for the "small TOC" addressing model.  Hence, as a precaution, use the "big TOC" model for compilation units that are intended for packing.

Basically, it's a heuristic: packed -> big TOC is probably necessary; not packed -> small TOC is generally enough.

**Comment by mshinwell** [comment]
@dra27 If you can approve the testsuite changes, this should be good to merge when CI passes.

**Review by dra27 — APPROVED** [review]
In general, it’s worth adding ocamltest actions in pairs (ie power and not_power) - but this is blocking other things, so don’t worry!
